model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/role/JpaRoleMapKeycloakTransaction.java

```diff
@@ -16,146 +16,51 @@
  */
 package org.keycloak.models.map.storage.jpa.role;
 
-import java.util.LinkedList;
-import java.util.List;
-import java.util.UUID;
-import java.util.stream.Stream;
 import javax.persistence.EntityManager;
 import javax.persistence.criteria.CriteriaBuilder;
-import javax.persistence.criteria.CriteriaDelete;
-import javax.persistence.criteria.CriteriaQuery;
-import javax.persistence.criteria.Order;
 import javax.persistence.criteria.Root;
-import org.keycloak.connections.jpa.JpaKeycloakTransaction;
+import javax.persistence.criteria.Selection;
 import org.keycloak.models.RoleModel;
-import static org.keycloak.models.jpa.PaginationUtils.paginateQuery;
-import org.keycloak.models.map.common.StringKeyConvertor.UUIDKey;
 import org.keycloak.models.map.role.MapRoleEntity;
 import org.keycloak.models.map.role.MapRoleEntityDelegate;
-import org.keycloak.models.map.storage.MapKeycloakTransaction;
-import org.keycloak.models.map.storage.QueryParameters;
-import static org.keycloak.models.map.storage.jpa.Constants.SUPPORTED_VERSION_ROLE;
-import static org.keycloak.models.map.storage.jpa.JpaMapStorageProviderFactory.CLONER;
+import static org.keycloak.models.map.storage.jpa.Constants.CURRENT_SCHEMA_VERSION_ROLE;
+import org.keycloak.models.map.storage.jpa.JpaMapKeycloakTransaction;
+import org.keycloak.models.map.storage.jpa.JpaModelCriteriaBuilder;
+import org.keycloak.models.map.storage.jpa.JpaRootEntity;
 import org.keycloak.models.map.storage.jpa.role.delegate.JpaRoleDelegateProvider;
 import org.keycloak.models.map.storage.jpa.role.entity.JpaRoleEntity;
-import static org.keycloak.utils.StreamsUtil.closing;
 
-public class JpaRoleMapKeycloakTransaction extends JpaKeycloakTransaction implements MapKeycloakTransaction<MapRoleEntity, RoleModel> {
+public class JpaRoleMapKeycloakTransaction extends JpaMapKeycloakTransaction<JpaRoleEntity, MapRoleEntity, RoleModel> {
 
+    @SuppressWarnings("unchecked")
     public JpaRoleMapKeycloakTransaction(EntityManager em) {
-        super(em);
+        super(JpaRoleEntity.class, em);
     }
 
     @Override
-    public MapRoleEntity create(MapRoleEntity mapEntity) {
-        JpaRoleEntity jpaEntity = (JpaRoleEntity) CLONER.from(mapEntity);
-        if (mapEntity.getId() == null) {
-            jpaEntity.setId(UUIDKey.INSTANCE.yieldNewUniqueKey().toString());
-        }
-        jpaEntity.setEntityVersion(SUPPORTED_VERSION_ROLE);
-        em.persist(jpaEntity);
-        return jpaEntity;
+    public Selection<JpaRoleEntity> selectCbConstruct(CriteriaBuilder cb, Root<JpaRoleEntity> root) {
+        return cb.construct(JpaRoleEntity.class, 
+            root.get("id"), 
```

Comment:
Adding column version for optimistic locking

```suggestion
            root.get("id"),
            root.get("version"),
```

Reply:
thanks, resolved.